cls/TestCoverage/Data/CodeUnit.cls

```diff
@@ -110,56 +125,95 @@ ClassMethod GetCurrentByName(pInternalName As %String, pSourceNamespace As %Stri
 		Set pCodeUnit.Type = tType
 		Set pCodeUnit.Hash = tHash
 		
-		Set tBitString = ""
-		For tLine=1:1:$Get(tCodeArray,0) {
-			Set $Bit(tBitString,tLine) = $Get(tExecutableFlags(tLine),0)
-		}
-		Set pCodeUnit.ExecutableLines = tBitString
+		
 		
 		If (tType = "CLS") {
 			Set pCodeUnit.Generated = ($$$comClassKeyGet(tName,$$$cCLASSgeneratedby) '= "")
 		}
 		
-		Set tMethod = ""
-		Set tMethodSignature = ""
-		Set tMethodMask = ""
-		For tLineNumber=1:1:$Get(tCodeArray,0) {
-			Set tLine = tCodeArray(tLineNumber)
-			Do pCodeUnit.Lines.Insert(tLine)
-			
-			If (tType = "CLS") {
-				// Extract line offset of methods in classes
-				Set tStart = $Piece(tLine," ")
-				If (tStart = "ClassMethod") || (tStart = "Method") {
-					Set tMethod = $Piece($Piece(tLine,"(")," ",2)
-					Set tMethodSignature = tLine
-					Do pCodeUnit.MethodMap.SetAt(tLineNumber,tMethod)
-					Do pCodeUnit.LineToMethodMap.SetAt(tMethod,tLineNumber)
-				} ElseIf ($Extract(tStart) = "{") {
-					// Ignore the opening bracket for a method.
-				} ElseIf ($Extract(tStart) = "}") && (tMethod '= "") {
-					// End of method. Add method subunit to class.
-					Set tSubUnit = ##class(TestCoverage.Data.CodeSubUnit.Method).%New()
-					Set tSubUnit.Name = tMethod
-					Set tSubUnit.DisplaySignature = tMethodSignature
-					Set tSubUnit.Mask = tMethodMask
-					Do pCodeUnit.SubUnits.Insert(tSubUnit)
-					Set tMethod = ""
-					Set tMethodSignature = ""
-					Set tMethodMask = ""
-				} ElseIf (tMethod '= "") {
-					Set $Bit(tMethodMask,tLineNumber) = 1
-				}
-			} Else {
-				// Extract line offset of labels in routines
-				If ($ZStrip($Extract(tLine),"*PWC") '= "") {
-					Set tLabel = $Piece($Piece(tLine," "),"(")
-					Do pCodeUnit.MethodMap.SetAt(tLineNumber,tLabel)
-					Do pCodeUnit.LineToMethodMap.SetAt(tLabel,tLineNumber)
+		
+		If (tType = "PY") {
+			// fill in the Lines property of this CodeUnit
+			set tPointer = 0
+			While $ListNext(pDocumentText, tPointer, tCurLine) {
+				do pCodeUnit.Lines.Insert(tCurLine)
+			}
+
+			do pCodeUnit.Lines.Insert("")
+
+			// Filling in the MethodMap and LineToMethodMap properties 
+			Set ClassName = $Piece(tName,".", *)
+			Set tMethodInfo = ##class(TestCoverage.Utils).GetPythonMethodMapping(pDocumentText, ClassName)
+			// tMethodInfo is a python tuple of (line to method info, method map info) 
+			Set tLineToMethodInfo = tMethodInfo."__getitem__"(0) // a python builtins list where the item at index i is the name of the method that line i is a part of 
+			Set tMethodMapInfo = tMethodInfo."__getitem__"(1) // a python builtins dict with key = method name, value = the line number of its definition
+			for i=1:1:$listlength(pDocumentText) {
+				Set tMethod = tLineToMethodInfo."__getitem__"(i)
+				Do pCodeUnit.LineToMethodMap.SetAt(tMethod,i)
+			}
+			Set iterator = tMethodMapInfo."__iter__"()
+			for i=1:1:tMethodMapInfo."__len__"() {
+				Set tMethod = iterator."__next__"()
```

Comment:
This feels like odd use of an iterator - wondering if there's a better pattern than iterating up to `__len__`

Reply:
I couldn't find anything much cleaner: if you do iterator."__next__" again after the last element, you get a Python Exception: StopIteration error, which I think is worse than the current method.